Add contextually aware title for block mover control (#557)

njpanderson opened the pull request:
After thinking about what would be best for the user in terms of accessibility and guidance, I’ve decided that trying to give them the “content” of a block wouldn’t very useful in a lot of contexts – especially non-text blocks.

Instead, I’ve opted for telling them where the block is, and where it’ll be moved to.

This is also the case for multiple selected blocks, where it will fall back to simpler language.

njpanderson committed:
Add contextually aware title block mover (#557)

After thinking about what would be best for the user in terms of accessibility and guidance, I’ve decided that trying to give them the “content” of a block wouldn’t very useful in a lot of contexts – especially non-text blocks.

Instead, I’ve opted for telling them where the block is, and where it’ll be moved to.

This is also the case for multiple selected blocks, where it will fall back to simpler language.

jasmussen commented:
Thanks for working on this. I see no visual change, so I'm assuming it's for screenreaders. Which is ⭐️ .

mtias commented on editor/block-mover/index.js:
```diff
@@ -13,26 +13,47 @@ import { IconButton } from 'components';
+import { getBlockType } from 'blocks';
 
-function BlockMover( { onMoveUp, onMoveDown, isFirst, isLast } ) {
+function BlockMover( { onMoveUp, onMoveDown, isFirst, isLast, uids, block, firstPosition } ) {
```
Why are both `firstPosition` and `isFirst` needed?

mtias commented on editor/block-mover/index.js:
```diff
@@ -13,26 +13,47 @@ import { IconButton } from 'components';
+	let blockType;
+
+	if ( uids.length === 1 && block ) {
+		blockType = getBlockType( block.name );
```
If the only thing you are using `blockType` for is the title, it'd make sense to move this to the `connect` function and assign a `blockTitle` prop there.

mtias commented on editor/block-mover/mover-label.js:
```diff
@@ -0,0 +1,73 @@
+import { __, sprintf } from 'i18n';
```
Needs to have the dependencies header comments.

mtias commented on editor/block-mover/mover-label.js:
```diff
@@ -0,0 +1,73 @@
+import { __, sprintf } from 'i18n';
+
+export function blockMoverLabel( selectedCount, { type, firstPosition, isFirst, isLast, dir } ) {
```
It'd be good to add a function comment here describing purpose and spec. Also, maybe `getBlockMoverLabel` is a better name.

njpanderson committed:
Rename firstPosition to firstIndex

njpanderson committed:
Move blockTitle logic into connect

njpanderson committed:
Move getBlockType import into correct chunk

njpanderson committed:
Add documentation for blockMoverLabel, etc

njpanderson committed:
Update multi/blockMoverLabel with ‘get’ prefix

njpanderson committed:
Fix test objects to use updated property

njpanderson commented:
@mtias - All above requests should have been handled in the most recent commits.

To answer the more specific queries:

> Why are both `firstPosition` and `isFirst` needed?

`firstPosition` is an index integer whereas `isFirst` is an existing boolean defining whether the block is first. I've renamed `firstPosition` to make clearer as to its use.

> If the only thing you are using `blockType` for is the title, it'd make sense to move this to the `connect` function and assign a `blockTitle` prop there.

This is done, although I haven't extended to the `.title` property of the resulting instance of `blockTitle` as the syntax looked a bit unwieldy in the context of the connect arguments. Happy to change if you think it's worth doing though?

> It'd be good to add a function comment here describing purpose and spec. Also, maybe `getBlockMoverLabel` is a better name.

Absolutely, and in my haste to get this over I'd completely forgotten about docs. Have added full JSDoc tags - although I'm not sure where you guys put new typedefs, so I've left them at the top of the `mover-label.js` file. Also renamed the functions with a `get` prefix.

njpanderson commented:
@jasmussen It is indeed! I've tested this with Voiceover and it works well. The change is intended to alter the `aria-label` property of the mover buttons.

njpanderson commented:
Just reminding people that I don't have merge abilities here - This code looks to be in a high traffic area of the repo so I'd be grateful if someone could re-review after the recent updates. Not sure how you manage this so apologies if I'm stepping on toes!

njpanderson committed:
Remove empty return value to avoid empty prop.

As suggested by @aduth in a previous PR, I’ve removed this (possibly unreachable) return and if none of the above cases are met, undefined will return which React can use to send an non-value prop, avoiding an empty aria-label attribute.

aduth commented on editor/block-mover/mover-label.js:
```diff
@@ -0,0 +1,108 @@
+
+	if ( dir < 0 && ! isFirst ) {
+		return sprintf(
+			__( 'Move %s blocks from position %s up by one place' ),
```
To accommodate languages which might swap placeholder ordering, it's recommended to use either [argument swapping](https://github.com/alexei/sprintf.js#argument-swapping) or [named placeholders](https://github.com/alexei/sprintf.js#named-arguments).

>If you have more than one placeholder in a string, it is recommended that you use argument swapping.

https://codex.wordpress.org/I18n_for_WordPress_Developers#Placeholders

There's a few other instances of this in the file.

Examples:

```js
__( 'Move %1$s blocks from position %2$s up by one place' ),
```
```js
__( 'Move %(selectedCount)s blocks from position %(position)s up by one place' ),
```

Additionally, a [`%d` format type](https://github.com/alexei/sprintf.js#format-specification) (numeric) might be the more applicable type to use here.

aduth commented on editor/block-mover/mover-label.js:
```diff
@@ -0,0 +1,108 @@
+ * @param  {blockMoverLabelOptions} options       Object options.
+ * @return {string}                               Label for the block movement controls.
+ */
+export function getBlockMoverLabel( selectedCount, { type, firstIndex, isFirst, isLast, dir } ) {
```
I'm not sure I understand the distinction between `selectedCount` as a standalone argument vs. `blockMoverLabelOptions` object. Why not have these all as named arguments `( selectedCount, type, firstIndex, isFirst, isLast, dir )` or all as object arguments `( { selectedCount, type, firstIndex, isLast, dir } )`?

aduth commented on editor/block-mover/test/mover-label.js:
```diff
@@ -0,0 +1,121 @@
+/**
```
Nice work on the tests 👍

njpanderson replied:
Thanks. Hopefully that's all the coverage you'd need.

njpanderson committed:
Convert second argument to named arguments

As suggested by @aduth, I’ve converted the second argument in both functions to separate named arguments. I’m not as keen on this signature but it makes for much tidier documentation (as well as removing the need for custom type definitions).

njpanderson committed:
Improve translation placeholders, add notes

njpanderson commented:
That's those updates made, @aduth. Let me know what you think.

aduth commented on editor/block-mover/mover-label.js:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	if ( isFirst && isLast ) {
+		// translators: %s: Type of block (i.e. Text, Image etc)
```
Looking at the output of `languages/gutenberg.pot` after running `npm run gettext-strings`, it appears these comments aren't being picked up. Running the code through [ASTExplorer](http://astexplorer.net/#/gist/d6fd6d8f9ec27a7312b135abfae973d8/5a1234076867231ab3657518682715d2e75f2208), it seems to be an issue that the comment isn't considered a `leadingComments` for the actual translation call. It can be fixed by the following:

```js
return sprintf( 
  /* translators: %s: Type of block (i.e. Text, Image etc) */
  __( 'Block "%s" is at the end of the content and can’t be moved down' ),
  type
);
```

However, I don't necessarily think you ought to need to make this change. I'd prefer if we could instead improve the tooling to be a bit more liberal in what it considers to be a leading comment for a translation, even if the actual function call occurs within `sprintf` or is the `return` value.

njpanderson replied:
Have left as-is for now, but if it is decided that the comment should need updated, it's a tiny change and I'm happy to make it.

aduth commented on editor/block-mover/test/mover-label.js:
```diff
@@ -0,0 +1,133 @@
+import { getBlockMoverLabel, getMultiBlockMoverLabel } from '../mover-label';
+
+describe( 'block mover', () => {
+	const dir_up = -1,
```
Variables should be named as camelCase, per guidelines:

https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#naming-conventions

Not sure if we really need the variables at all really, or at least not exclusively in the context of tests, since it doesn't really do anything to safeguard us against refactoring of the code under test.

njpanderson replied:
Ugh, sorry about that @aduth - In my defence this is my first ever WP contribution! Will read up on the standards and get better acquainted with the way things are done here.

On another note, I see eslint does have camelCase enforcement built in and wonder if we'd benefit from putting it into the `.eslintrc` file? Happy to do another PR if you'd like to add this.

njpanderson replied:
Just to clarify - the variables `dirUp` and `dirDown` exist purely to guide future testers as to what the last argument is for. One could argue "why not just have all the arguments named", but the others felt too varied to be worth being named. Am happy to just remove the vars though, if you feel it would be more consistent with the rest of the test suites.

aduth replied:
My worry is: Why is this relevant to the tests only, and not runtime usage?

The alternative I see is exporting constants from `mover-label.js`:

```js
export const DIRECTION_UP = -1;
export const DIRECTION_DOWN = 1;
```

_(Realizing now that JavaScript standards have no equivalent to [PHP's upper-case constants](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#naming-conventions))_

These constants could be imported both in tests, but also in `<BlockMover />`, and can be revised once and apply everywhere.

But still, I think it might be overkill. I'm not really requesting any changes here, just trying to understand why we'd care to special-case the tests.

njpanderson replied:
> My worry is: Why is this relevant to the tests only, and not runtime usage?

For runtime usage, it's only invoked from two statements in `block-mover/index.js` - the only reason I added these variables was to avoid the repetition of literals specifically within the tests. You're right though, when other arguments are still static it feels a bit odd.

I'm not massively wedded to it at all and am more than happy just to submit another PR removing them.

You might be onto something with the direction constants though — I could, however, make them static properties of the `getBlockMoverLabel` function:

```
function getBlockMoverLabel() { ... }

getBlockMoverLabel.DIRECTION_UP = -1;
getBlockMoverLabel.DIRECTION_DOWN = 1;

export { getBlockMoverLabel };
```

Happy to do something like that if you'd like, assuming there's no issue with the constant syntax suggested above.

aduth reviewed (changes requested):
With variable rename, I think this should be good to go 👍

njpanderson committed:
Update case of variables
